controllers/ControllerDecorator.ts

```diff
@@ -42,32 +88,112 @@ export function registerControllers(baseUrl: string = '', router: Router = Route
+                routeId = routeUrl + route.method.toString(),
+                returnType = Reflect.getMetadata('design:returntype', routeTarget, route.propertyKey),
+                params: Param[] = Reflect.getOwnMetadata(PARAMS_KEY, routeTarget, route.propertyKey) || [],
+                method = route.descriptor.value,
+                hasResponseParam = !!params.filter(p => p.paramType === ParamType.Response).length;
+
+            if (definedRoutes.indexOf(routeId) !== -1) {
+                throw new DuplicateRouteDeclarationError(routeUrl, route.method);
+            }
+
+            params.forEach(p => {
+                if (p.type.length < 1) {
+                    throw new ParameterConstructorArgumentsError(p.name);
+                }
+            });
 
             if (routeUrl.length > 1 && routeUrl[routeUrl.length - 1] === '/') {
                 routeUrl = routeUrl.substr(0, routeUrl.length - 1);
             }
 
-            if (definedRoutes.indexOf(routeId) !== -1) {
-                throw new DuplicateRouteDeclarationError(routeUrl, route.method);
-            }
+            params = params.sort((p1, p2) => (p1.index < p2.index) ? -1 : 1);
+
+            route.descriptor.value = (request: Request, response: Response, next) => {
+                let errorHandlers: ErrorHandlerManager = Reflect.getMetadata(ERRORHANDLER_KEY, ctrlTarget),
+                    paramValues = [];
+
+                if (!errorHandlers) {
+                    errorHandlers = new ErrorHandlerManager();
+                    errorHandlers.addHandler(defaultErrorHandler);
+                }
+
+                try {
+                    params.forEach((p: Param) => {
+                        switch (p.paramType) {
+                            case ParamType.Request:
+                                paramValues[p.index] = request;
+                                return;
+                            case ParamType.Response:
+                                paramValues[p.index] = response;
+                                return;
+                            case ParamType.Body:
+                                paramValues[p.index] = parseParam(request.body, p);
+                                return;
+                            case ParamType.Url:
+                                paramValues[p.index] = parseParam(request.params[p.name], p);
+                                return;
+                            case ParamType.Query:
+                                paramValues[p.index] = parseParam(request.query[p.name], p);
+                                return;
+                            case ParamType.Header:
+                                paramValues[p.index] = parseParam(request.get(p.name), p);
+                                return;
+                        }
+                    });
+                } catch (e) {
+                    errorHandlers.callHandlers(request, response, e);
+                    return;
+                }
```

Comment:
I would extract this whole try-catch-block to a buildParamValues-function. Just for readability.